File: solutions/go/phone-number/1/phone_number.go
package phonenumber

import (
	"errors"
	"regexp"
)

func Number(phoneNumber string) (string, error) {
	if phoneNumber == "" {
		return "", errors.New("")
	}

	reg := regexp.MustCompile("([^0-9]+)")
	phoneNumber = reg.ReplaceAllString(phoneNumber, "")

	if len(phoneNumber) == 11 {
		if phoneNumber[0] == '1' {
			phoneNumber = phoneNumber[1:]
			if phoneNumber[0] == '0' || phoneNumber[0] == '1' || phoneNumber[3] == '0' || phoneNumber[3] == '1' {
				return "", errors.New("")
			}
			return phoneNumber, nil
		}
	} else if len(phoneNumber) == 10 {
		if phoneNumber[0] == '0' || phoneNumber[0] == '1' || phoneNumber[3] == '0' || phoneNumber[3] == '1' {
			return "", errors.New("")
		}
		return phoneNumber, nil
	}

	return "", errors.New("")
}

func AreaCode(phoneNumber string) (string, error) {
	result, err := Number(phoneNumber)
	if err == nil {
		return result[0:3], err
	}
	return "", err
}

func Format(phoneNumber string) (string, error) {
	result, err := Number(phoneNumber)
	if err != nil {
		return "", errors.New("")
	}

	result = "(" + result[0:3] + ") " + result[3:6] + "-" + result[6:]
	return result, nil
}

File: solutions/cpp/phone-number/1/phone_number.cpp
#include "phone_number.h"
#include <algorithm>
#include <cctype>
#include <pstl/glue_algorithm_defs.h>
#include <stdexcept>
#include <string>

namespace phone_number {
phone_number::phone_number(const std::string &input) : num(input) {
  num.erase(std::remove_if(num.begin(), num.end(),
                           [](const char &c) {
                             if (std::isdigit(c)) {
                               return false;
                             } else {
                               return true;
                             }
                           }),
            num.end());
  if (num.length() <= 9 || num.length() > 11) {
    throw std::domain_error("");
  }
  if (num.length() == 11) {
    if (num.at(0) == '1') {
      num.erase(0, 1);
    } else {
      throw std::domain_error("");
    }
  }
  if (num.at(0) == '0' || num.at(0) == '1' || num.at(3) == '0' ||
      num.at(3) == '1')
    throw std::domain_error("");
}

const std::string phone_number::number() const { return num; }

const std::string phone_number::area_code() const {
  const std::string result = num.substr(0, 3);
  return result;
}

phone_number::operator std::string() const {
  std::string result =
      "(" + area_code() + ") " + num.substr(3, 3) + "-" + num.substr(6);
  return result;
}

} // namespace phone_number
